chain/reorg_test.go

```diff
@@ -1,71 +1,141 @@
 	th "github.com/filecoin-project/go-filecoin/testhelpers"
 	tf "github.com/filecoin-project/go-filecoin/testhelpers/testflags"
 	"github.com/filecoin-project/go-filecoin/types"
 )
 
-func TestIsReorg(t *testing.T) {
+func TestIsReorgFork(t *testing.T) {
 	tf.UnitTest(t)
+	dstP := initDSTParams()
+	ctx, blockSource, chainStore := setupGetAncestorTests(t, dstP)
+	old, new, common := getForkOldNewCommon(ctx, t, chainStore, blockSource, dstP)
+	assert.True(t, chain.IsReorg(old, new, common))
+}
+func TestIsReorgPrefix(t *testing.T) {
+	tf.UnitTest(t)
+	dstP := initDSTParams()
+	ctx, blockSource, chainStore := setupGetAncestorTests(t, dstP)
+	old, new, common := getPrefixOldNewCommon(ctx, t, chainStore, blockSource, dstP)
+	assert.False(t, chain.IsReorg(old, new, common))
+}
+
+func TestIsReorgSubset(t *testing.T) {
+	tf.UnitTest(t)
+	dstP := initDSTParams()
+	ctx, blockSource, chainStore := setupGetAncestorTests(t, dstP)
+	old, new, common := getSubsetOldNewCommon(ctx, t, chainStore, blockSource, dstP)
+	assert.False(t, chain.IsReorg(old, new, common))
+}
+
+func TestReorgDiffFork(t *testing.T) {
+	tf.UnitTest(t)
+	dstP := initDSTParams()
+	ctx, blockSource, chainStore := setupGetAncestorTests(t, dstP)
+	old, new, common := getForkOldNewCommon(ctx, t, chainStore, blockSource, dstP)
+
+	dropped, added, err := chain.ReorgDiff(old, new, common)
+	assert.NoError(t, err)
+	assert.Equal(t, uint64(11), dropped)
+	assert.Equal(t, uint64(14), added)
+}
+
+func TestReorgDiffSubset(t *testing.T) {
+	tf.UnitTest(t)
+	dstP := initDSTParams()
+	ctx, blockSource, chainStore := setupGetAncestorTests(t, dstP)
+	old, new, common := getSubsetOldNewCommon(ctx, t, chainStore, blockSource, dstP)
+
+	dropped, added, err := chain.ReorgDiff(old, new, common)
+	assert.NoError(t, err)
+	assert.Equal(t, uint64(1), dropped)
+	assert.Equal(t, uint64(1), added)
+}
+
+// getForkOldNewCommon
+func getForkOldNewCommon(ctx context.Context, t *testing.T, chainStore chain.Store, blockSource *th.TestFetcher, dstP *DefaultSyncerTestParams) (types.TipSet, types.TipSet, types.TipSet) {
+	// Add 10 tipsets to the head of the chainStore.
+	requireGrowChain(ctx, t, blockSource, chainStore, 10, dstP)
+	commonAncestor := requireHeadTipset(t, chainStore)
+
+	// make the first fork tipset
+	signer, ki := types.NewMockSignersAndKeyInfo(1)
+	mockSignerPubKey := ki[0].PublicKey()
+	fakeChildParams := th.FakeChildParams{
+		Parent:      commonAncestor,
+		GenesisCid:  dstP.genCid,
+		Signer:      signer,
+		MinerPubKey: mockSignerPubKey,
+		StateRoot:   dstP.genStateRoot,
+		Nonce:       uint64(4),
+	}
+
+	firstForkBlock := th.RequireMkFakeChild(t, fakeChildParams)
+	requirePutBlocks(t, blockSource, firstForkBlock)
+	firstForkTS := th.RequireNewTipSet(t, firstForkBlock)
+	firstForkTsas := &chain.TipSetAndState{
+		TipSet:          firstForkTS,
+		TipSetStateRoot: dstP.genStateRoot,
+	}
+	th.RequirePutTsas(ctx, t, chainStore, firstForkTsas)
+	err := chainStore.SetHead(ctx, firstForkTS)
```

Comment:
Does setting the head here control which chain `requireGrowChain` grows?

Reply:
Yes
General note: the testing framework that has grown organically in this package is in need of some intentional redesign.